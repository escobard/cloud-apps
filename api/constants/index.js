const dbkeys = require("./dbkeys"),
  routes = require("./routes")

module.exports = {
  dbkeys,
  routes
}
