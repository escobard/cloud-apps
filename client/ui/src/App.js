import React from "react";

import { BrowserRouter as Router, Switch, Route } from "react-router-dom";

import { Home } from "pages";

import { Header, Footer } from "components";

import "styles/global.scss";

const App = () => {
  return (
    <Router>
      <>
        <Header />
        <div className="divider" />
        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route path="*">
            <span>Page not found.</span>
          </Route>
        </Switch>
        <Footer />
      </>
    </Router>
  );
};

export default App;
