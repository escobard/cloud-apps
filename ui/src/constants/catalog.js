import { todayDate } from "utils/todayDate";

export const footer = {
  noNotes: "NO NOTES",
  withNotes: "COMPLETED"
};

export const header = {
  title: "Your Notes",
  todayDate: todayDate()
};

export const notes = {
  noNotes: {
    subject: "No Notes",
    note: "Add a note by clicking on the + icon below.",
    icon: {
      image: "exclamation",
      label: "No Notes"
    }
  },
  apiError: {
    subject: "API error",
    icon: {
      image: "exclamation",
      label: "API error"
    }
  }
};
