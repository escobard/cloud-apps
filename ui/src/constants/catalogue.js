import { todayDate } from "utils/todayDate";

export const footer = {
  noNotes: "NO NOTES",
  withNotes: "COMPLETED"
};

export const header = {
  title: "Your Notes",
  todayDate: todayDate()
}