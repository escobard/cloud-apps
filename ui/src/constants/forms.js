export const addNoteFields = [
  {
    name: "subject",
    label: "Subject",
    placeholder: "Enter a subject for your note",
    value: "",
    errors: {
      // expand this later with future error strings
    }
  },
  {
    name: "note",
    label: "Note *",
    placeholder: "Enter a description for your note (required)",
    value: "",
    errors: {}
  }
];