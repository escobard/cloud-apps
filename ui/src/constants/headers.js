export const headers = { "Access-Control-Allow-Origin": "*" };
