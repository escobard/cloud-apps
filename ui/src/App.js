import React, { Component, Fragment } from "react";

import Header from "./components/Header";
import Form from "./components/Form";
import Footer from "./components/Footer";
import Note from "./components/Note";
import Modal from "./components/Modal";

import { addNoteFields } from "./constants";
import { addNote, getNotes } from "./utils/requests";

import "./styles/global.scss";

class App extends Component {
  state = {
    messageErrors: [],
    title: "Post Form",
    message:
      "Follow the placeholder instructions to validate data on the UI and API side",
    status: false,
    show: false,
    notes: []
  };

  async componentDidMount(){
    const notes = await getNotes();
    this.setState({notes})
  }

  open = size => () => this.setState({ size, show: true });
  close = () => this.setState({ show: false });

  /** Submits the POST request to the API
   * @name addNote
   * @dev this requests tests basic validation between UI and API
   * @param {string} subject, contains note's subject value
   * @param {string} note, contains note's note value
   * @returns /addNote route response, or validation message
   **/

  addNote = async (subject, note) => {
    let { messageErrors } = this.state;

    // triggers validation logic
    this.validateAddNote(subject, note);

    // only runs request, if no validation message are present
    if (messageErrors.length === 0) {
      const request = {
        subject,
        note
      };

      let response = await addNote(request);

      // checks for API promise rejections
      if (!response.status) {
        return this.setState({
          title: "addNote() error(s)",
          message: response,
          status: "red"
        });
      } else if (response.data.note) {
        const {
          data: { status }
        } = response;

        this.setState({
          title: "addNote() added!",
          message: status,
          status: "green"
        });

        const notes = await getNotes()

        // add timeout here to close out modal on note creation
        setTimeout(async () =>{
          this.setState({notes})
          this.close()
        }, 500 )
      }
    }
  };

  /** Resets the message array after form validation checks
   * @returns this.setState()
   **/

  emptyErrors = () => {
    this.setState({
      messageErrors: []
    });
  };

  /** Validates a form value
   * @dev can be split out into a validation class to re-use in api / ui layers
   * @param {*} value, property to validate
   * @param {*} condition, functional condition to validate / invalidate value
   * @param {string} error, string of error to add to this.state.message
   **/

  validateField = (value, condition, error) => {
    if (condition) {
      this.setState({ messageErrors: this.state.messageErrors.push(error) });
    }
  };

  /** Validates addNote values
   * @name validateAddNote
   * @dev used to reduce clutter in makeDonation
   * @param {string} subject, contains random string value
   * @param {string} note, contains random string value with a length greater than 10
   **/

  validateAddNote = (subject, note) => {
    let { messageErrors } = this.state;

    this.validateField(
      note,
      note.length < 25,
      "Note must contain more than 25 characters"
    );

    // sets messagesState
    if (messageErrors.length > 0) {
      this.setState({
        status: "red",
        title: "addNote() form error:",
        message: `Form contains the following error: ${messageErrors.join(
          ", "
        )}.`
      });
      this.emptyErrors();
    } else {
      this.setState({
        status: "green",
        title: "addNote() validated",
        message: `Adding note...`
      });
    }
  };

   /** Renders Notes based on API response
   * @name renderNotes
   * @dev used to render multiple notes
   * @param {array} data, contains random string value
   * @returns one or more <Note />
   **/

  renderNotes(id, data){
    if (data.length === 0) {return <p>No notes</p>}
    return data.map((object, index) =>{
      return <Note key={id + index} id={`${id}-${index}`} data={object} />
    })
  }

  render() {
    const id = "application";

    let {
      title,
      message,
      status,
      show,
      notes
    } = this.state;

    const note = {
      title: "Test title",
      note:
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis vel nulla sit amet nibh sagittis eleifend. Cras a lacus rutrum ipsum pretium scelerisque sed eu turpis. ",
      date: "9 am"
    };

    const formMessage =
      status
        ? {
            color: status,
            header: title,
            content: message
          }
        : null;

    return (
      <Fragment>
        <Header />
        <div className="divider" />
        <main id={id} className="application">
          <Modal
            id={id}
            title="Add Note"
            open={show}
            close={this.close}
            content={
              <Form
                id={id}
                message={formMessage}
                addNote={this.addNote}
                fields={addNoteFields}
              />
            }
          />
          {this.renderNotes(id, notes)}
          <Footer id={id} open={this.open()} />
        </main>
      </Fragment>
    );
  }
}

export default App;
