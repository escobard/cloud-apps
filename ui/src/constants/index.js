import { addNoteFields } from "./forms";
import { apiRoutes } from "./urls";
import { headers } from "./headers";
import { notes, footer, header } from "./catalog";

export { apiRoutes, addNoteFields, headers, notes, footer, header };
