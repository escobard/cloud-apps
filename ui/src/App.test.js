import { notes } from "./constants/catalog";
import App from "./App";

describe("App", () => {
  afterEach(() => {
    cleanup();
  });
  const response = {
    data: [
      {
        id: 1,
        user_id: 1,
        subject: "test subject",
        note: "test note",
        date: "03/22/20, 10:23 PM"
      }
    ]
  };
  describe(">> addNote()", () => {
    it(">> displays a validation form error", () => {});
    it(">> user can send a note", () => {});
  });

  describe(">> renderNotes()", () => {
    it(">> renders notes in initial state", () => {
      const { getByText } = render(<App />);
      expect(getByText(notes.noNotes.subject));
    });

    it(">> renders notes in error state", async () => {
      axios.get.mockRejectedValue("TEST");
      const { rerender, getByText } = render(<App />);
      await act(async () => {
        rerender(<App />);
      });
      expect(await getByText(notes.apiError.subject));
    });

    it(">> renders notes with data state", async () => {
      axios.get.mockResolvedValue(response);
      const { rerender, getByText } = render(<App />);
      await act(async () => {
        rerender(<App />);
      });
      expect(await getByText(response.data[0].subject));
    });
  });

  describe(">> render()", () => {
    it(">> snapshot is up to date", () => {
      const { container } = render(<App />);
      expect(container).toMatchSnapshot();
    });
  });
});
