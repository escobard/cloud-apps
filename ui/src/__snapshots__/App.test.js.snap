// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App >> render() >> snapshot is up to date 1`] = `
<div>
  <nav
    id="application-navigation"
  >
    <i
      aria-label="Menu"
      class="bars big icon"
    />
    <h1>
      Your Notes
    </h1>
    <p>
      March 23, 2020
    </p>
  </nav>
  <div
    class="divider"
  />
  <main
    class="application"
    id="application"
  >
    <div
      class="ui equal width grid"
      id="application-no-notes-note"
    >
      <div
        class="one wide computer three wide mobile two wide tablet column icon"
      >
        <i
          aria-label="No Notes"
          class="exclamation big circular icon error"
        />
      </div>
      <div
        class="thirteen wide computer nine wide mobile twelve wide tablet column content"
      >
        <h3>
          No Notes
        </h3>
        <p>
          Add a note by clicking on the + icon below.
        </p>
      </div>
    </div>
  </main>
  <footer
    id="application-footer"
  >
    <p>
      NO NOTES
    </p>
    <i
      class="plus big icon"
      role="button"
    />
  </footer>
</div>
`;
